docs/design/rosbag2_snapshot_recorder/rosbag2_snapshot_recorder.md

```diff
@@ -0,0 +1,63 @@
+
+## Goal
+
+Rosbag in ROS1 had functionality of creating a [snapshot](https://github.com/ros/rosbag_snapshot). Snapshotter was a program which subscribed to topics and buffered messages until a `trigger` for snapshot was sent. As in Rosbag2 we want to support the following features:
+* Save data to rosbag on trigger
+  * Enable triggering  buffer save to disk with service call
+* Node and buffer configuration
+  * Enable user to specify `cli` arguments which could be shared with `ros2 bag record` / `Recorder` node or similar to them (dependent on design). User should be able to configure buffer size (maximum memory per topic) and duration (maximum time difference between newest and oldest message), subscribed topics (list of topics or all currently available topics) and output rosbag filename.
+* Pause and resume buffering process
+  * Pausing refers to stopping buffering new messages until resumed or write is triggered.
+  * Resuming continues the process of buffering after pause. Resuming would clear all buffers.
+
+
+## Proposal
+
+As thinking of this design in the past `rosbag_snapshot` was written as different package than `rosbag`. It was defined this way not to break the existent API of `rosbag`. Taken this into consideration it could be done so this time too. However it would require duplicating a large part of code which is already written and being optimized as a part of `rosbag2` repository. So as not being sure which way would be the most appropriate to implement I propose multiple approaches.
+
+### Outside rosbag2
+
+First approach implements `ros2 bag snapshot` as a different package. Mostly it requires rewriting `Recorder` node completly. Completly means in this case diving into implementation of `Writer`, `SequentialWriter` and `BaseWriterInterface` to copy and adjust writing to disk algorithms.
+
+Pros:
+* Design flexibility
+
+Cons:
+* Many new functions would be redefined or even redundant
+* Breaking the initial desing of writers/readers in favor of putting more effort to custom mechanism
+* Not being part of `rosbag2` and `ros2bag` (cli incompatibility)
+
+### Writer Interfaces changes + `Recorder` modifications
+
+Second approach would base on making changes to `Recorder`, writers and their interfaces. This approach bases on adding service to `Recorder` which handles triggering buffer save. Although current design of writers bases on `BaseWriterInterface` which would need to be changed to enable some kind of `save()/flush()` functionality. I believe this is not an elegant approach but I haven't find other way to trigger buffer save from `Recorder`, through `Writer` to `SequentialWriter`. For sure it requires implementing new `SnapshotWriter` since it needs to take care of buffering the messages from subscribers instead of passing them straight to `MessageCache` and `CacheConsumer`.
+
+Pros:
+* Maintaining snapshot inside `ros2 bag snapshot ...` (or even `ros2 bag record snapshot ...`) and `Recorder` functionality
```

Comment:
I'm not sure that this makes sense as a separate verb. I would think that the circular-buffer + snapshot approach should be considered a behavioral modification of recording, just like `rosbag2 play --loop`

I'd like to see a specific section for "User Interface" - how do I interact with this feature (without knowing how it is built)?

I think maybe something like the following:

```
# Recording is invoked with this feature via an extra option to "record"
ros2 bag record --buffer-snapshot

# Snapshot is exposed as a Service on the Recorder node (so it can be triggered remotely)
ros2 service call /rosbag2_recorder rosbag2_interfaces/Snapshot

# C++ API
* RecordingOptions has a new value "buffer_snapshot"
* Recorder has a new API `Recorder::snapshot`
```